### PR TITLE
Added textual inversion distillation for quantization example.

### PR DESCRIPTION
**Pull request by XinyuYe-Intel**
# What does this PR do?

Added a textual inversion distillation for quantization example.

## Before submitting
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

**Comment by hshen14** [comment]
@echarlaix  please help review. This could be the first time to generate a static INT8 model for fine-tuned SD.

File: examples/neural_compressor/textual_inversion/textual_inversion.py [hunk 1]
```diff
@@ -0,0 +1,894 @@
...
+            pipeline = pipeline.to(torch.device("cpu"))
+        
+        loaded_model_images = generate_images(pipeline, prompt=prompt, seed=args.seed)
+        if loaded_model_images != optimized_model_images:
+            logger.info("The quantized model was not successfully loaded.")
+        else:
+            logger.info(f"The quantized model was successfully loaded.")
+
+if __name__ == "__main__":
+    main()
```
Comment (echarlaix) [inline examples/neural_compressor/textual_inversion/textual_inversion.py:894]: renaming the directory to `"textual-inversion"` would be best to keep the same format

File: examples/neural_compressor/textual_inversion/text2images.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+unet = UNet2DConditionModel.from_pretrained(args.pretrained_model_name_or_path, subfolder="unet")
+
+pipeline = StableDiffusionPipeline(
+    text_encoder=text_encoder,
+    vae=vae,
+    unet=unet,
+    tokenizer=tokenizer,
+    scheduler=PNDMScheduler.from_config("CompVis/stable-diffusion-v1-4", subfolder="scheduler"),
+    safety_checker=StableDiffusionSafetyChecker.from_pretrained("CompVis/stable-diffusion-safety-checker"),
+    feature_extractor=CLIPFeatureExtractor.from_pretrained("openai/clip-vit-base-patch32"),
```
Comment (echarlaix) [inline examples/neural_compressor/textual_inversion/text2images.py:101]: Why is this hardcoded ?

File: examples/neural_compressor/textual_inversion/textual_inversion.py [hunk 3]
```diff
@@ -0,0 +1,894 @@
...
+    )
+    unet = UNet2DConditionModel.from_pretrained(
+        args.pretrained_model_name_or_path,
+        subfolder="unet",
+        revision=args.revision,
+    )
+
+    scheduler = PNDMScheduler.from_config("CompVis/stable-diffusion-v1-4", subfolder="scheduler")
+    safety_checker = StableDiffusionSafetyChecker.from_pretrained("CompVis/stable-diffusion-safety-checker")
+    feature_extractor = CLIPFeatureExtractor.from_pretrained("openai/clip-vit-base-patch32")
```
Comment (echarlaix) [inline examples/neural_compressor/textual_inversion/textual_inversion.py:521]: Why is this hardcoded ?

File: examples/neural_compressor/textual_inversion/README.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+Before running the scripts, make sure to install the library's training dependencies:
+
+```bash
+pip install -r requirements.txt
+```
+
+## Prepare Datasets
+
+One picture which is from the huggingface datasets [sd-concepts-library/dicoo2](https://huggingface.co/sd-concepts-library/dicoo2) is needed, and save it to the `./dicoo` directory. The picture show below:
```
Comment (echarlaix) [inline examples/neural_compressor/textual_inversion/README.md:17]: I would be in favor of not saving any images to not add weight to the repo

File: examples/neural_compressor/textual_inversion/README.md [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+## Do distillation for quantization
+
+Once you have the FP32 Textual Inversion model, the following command will take the FP32 Textual Inversion model as input to do distillation for quantization and generate the INT8 Textual Inversion model.
+
+```bash
+export FP32_MODEL_NAME="./dicoo_model"
+export DATA_DIR="./dicoo"
+
+accelerate launch textual_inversion.py \
```
Comment (echarlaix) [inline examples/neural_compressor/textual_inversion/README.md:54]: why use accelerate for this second step and not the first ?

File: examples/neural_compressor/textual_inversion/textual_inversion.py [hunk 6]
```diff
@@ -0,0 +1,894 @@
...
+        model = copy.deepcopy(unet)
+        confs = []
+        if args.do_quantization:                               
+            from neural_compressor import QuantizationAwareTrainingConfig
+            q_conf = QuantizationAwareTrainingConfig()
+            confs.append(q_conf)
+
+        if args.do_distillation:
+            teacher_model = copy.deepcopy(model)
+            layer_mappings=[
```
Comment (echarlaix) [inline examples/neural_compressor/textual_inversion/textual_inversion.py:762]: This will vary for different architectures types, so I'll be in favor of removing distillation here

File: examples/neural_compressor/textual_inversion/textual_inversion.py (hunk 3 above)
Comment (XinyuYe-Intel) [reply examples/neural_compressor/textual_inversion/textual_inversion.py:521]: It's originated from diffusers' example code.

File: examples/neural_compressor/textual_inversion/README.md (hunk 4 above)
Comment (XinyuYe-Intel) [reply examples/neural_compressor/textual_inversion/README.md:17]: A few images won't take too much space, while it helps demonstrate the example more vividly, so I think the benefit it brings may outweigh the downside of this.

File: examples/neural_compressor/textual_inversion/textual_inversion.py (hunk 6 above)
Comment (XinyuYe-Intel) [reply examples/neural_compressor/textual_inversion/textual_inversion.py:762]: Actually the distillation helps QAT a lot in terms of output quality of INT8 model, not sure removing it will bring the same result as this one.

File: examples/neural_compressor/textual_inversion/textual_inversion.py (hunk 1 above)
Comment (XinyuYe-Intel) [reply examples/neural_compressor/textual_inversion/textual_inversion.py:894]: will do.

File: examples/neural_compressor/textual_inversion/README.md (hunk 5 above)
Comment (XinyuYe-Intel) [reply examples/neural_compressor/textual_inversion/README.md:54]: will change the first one to use accelerate also.

File: examples/neural_compressor/textual_inversion/textual_inversion.py (hunk 3 above)
Comment (echarlaix) [reply examples/neural_compressor/textual_inversion/textual_inversion.py]: I don't see it in example [script](https://github.com/huggingface/diffusers/blob/v0.14.0/examples/textual_inversion/textual_inversion.py#L101) of the lastest release, could this be updated ?

**Review by echarlaix — COMMENTED** [review]
Can you run `make style` using `black==23.1.0` and `isort==5.10.1` to fix the code quality test ?

File: examples/neural_compressor/textual_inversion/textual_inversion.py (hunk 6 above)
Comment (XinyuYe-Intel) [reply examples/neural_compressor/textual_inversion/textual_inversion.py]: Sure, it's reasonable.

**Comment by XinyuYe-Intel** [comment]
> Can you run `make style` using `black==23.1.0` and `isort==5.10.1` to fix the code quality test ?



> Can you run `make style` using `black==23.1.0` and `isort==5.10.1` to fix the code quality test ?

Will do.

File: examples/neural_compressor/textual_inversion/textual_inversion.py (hunk 6 above)
Comment (echarlaix) [reply examples/neural_compressor/textual_inversion/textual_inversion.py]: Thanks a lot !

File: examples/neural_compressor/textual-inversion/text2images.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    text_encoder=text_encoder,
+    vae=vae,
+    unet=unet,
+    tokenizer=tokenizer,
+    scheduler=PNDMScheduler.from_config("CompVis/stable-diffusion-v1-4", subfolder="scheduler"),
+    safety_checker=StableDiffusionSafetyChecker.from_pretrained("CompVis/stable-diffusion-safety-checker"),
+    feature_extractor=CLIPFeatureExtractor.from_pretrained("openai/clip-vit-base-patch32"),
+)
+pipeline.safety_checker = lambda images, clip_input: (images, False)
+if os.path.exists(os.path.join(args.pretrained_model_name_or_path, "best_model.pt")):
```
[the remark this answers is not in the document]
Comment (XinyuYe-Intel) [reply examples/neural_compressor/textual-inversion/text2images.py:107]: This is due to the UNet is an INT8 model produced by INC's API, if we change the name, may not be able to load it with INC's API.

File: examples/neural_compressor/textual_inversion/textual_inversion.py (hunk 6 above)
Comment (XinyuYe-Intel) [reply examples/neural_compressor/textual_inversion/textual_inversion.py]: done.

File: examples/neural_compressor/textual_inversion/README.md (hunk 4 above)
Comment (XinyuYe-Intel) [reply examples/neural_compressor/textual_inversion/README.md]: changed images to links.

File: examples/neural_compressor/textual_inversion/textual_inversion.py (hunk 3 above)
Comment (XinyuYe-Intel) [reply examples/neural_compressor/textual_inversion/textual_inversion.py]: fixed.

**Comment by XinyuYe-Intel** [comment]
> Can you run `make style` using `black==23.1.0` and `isort==5.10.1` to fix the code quality test ?

I did `make style` using `black==23.1.0` and `isort==5.10.1`, but looks like issue is still there.

**Comment by echarlaix** [comment]
> I did `make style` using `black==23.1.0` and `isort==5.10.1`, but looks like issue is still there.

Thanks a lot for iterating on it @XinyuYe-Intel !! Could you rebase and run `pip install --upgrade -e .["quality"]` before running `make style` once again ?

File: examples/neural_compressor/textual-inversion/text2images.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (echarlaix) [reply examples/neural_compressor/textual-inversion/text2images.py:100]: We should be able as we do with the `INCModel` no ?

[the remark this answers is not in the document]
Comment (XinyuYe-Intel) [reply examples/neural_compressor/textual-inversion/text2images.py:99]: I think this may requires to implement a class called `INCUNet2DConditionModel`. By the way, if we save the model as `diffusion_pytorch_model.bin`, we cannot tell the model is a fp32 model or an int8 model unless user tells us, so we may need to add an argument like `--int8` in the `text2images.py`.

**Comment by XinyuYe-Intel** [comment]
> > I did `make style` using `black==23.1.0` and `isort==5.10.1`, but looks like issue is still there.
> 
> Thanks a lot for iterating on it @XinyuYe-Intel !! Could you rebase and run `pip install --upgrade -e .["quality"]` before running `make style` once again ?

No problem. I've done the above steps, but the code quality issue remains.
